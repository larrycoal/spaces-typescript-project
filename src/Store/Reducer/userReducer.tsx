import { Action } from '../Actions/types'
type User = { isLoggedIn: Boolean }



const userReducer = (state: User = { isLoggedIn: false }, action: Action) => {
    switch (action.type) {
        case "login": return {
            isLoggedIn: action.payload
        }
        case "fetchUser" : return state
        default: return state
    }
}

export default userReducer