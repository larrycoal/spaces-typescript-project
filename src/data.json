[
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "red"
    },
    {
        "name": "Lorem",
        "shape": "rectangle",
        "color": "green"
    },
    {
        "name": "Lorem",
        "shape": "square",
        "color": "red"
    },
    {
        "name": "Lorem",
        "shape": "round",
        "color": "red"
    },
    {
        "name": "Lorem",
        "shape": "triangle",
        "color": "red"
    },
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "blue"
    },
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "yellow"
    },
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "blue"
    },
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "red"
    },
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "grey"
    },
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "blue"
    },
    {
        "name": "Lorem",
        "shape": "oval",
        "color": "red"
    }
]