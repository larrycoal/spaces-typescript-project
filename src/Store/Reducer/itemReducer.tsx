//import { Action } from '../Actions/types'
import data from '../../data.json'

const initialState = {
    item:data,
    filterName:"All Items"
}

export interface itemState {
    item:{
        name:String,
        shape:String,
        color:String
    }[],
    filterName:String
}
interface Filter {
    type:String,
    payload: {shape:String[],color:String[]}
}


const filter = (filters:Filter["payload"]):itemState=>{
   let newData:itemState["item"] = data
    let {shape,color} =filters
   let filteredData:itemState["item"] = []
   let filteredTitle:itemState["filterName"]=""
  newData.map((data)=>{
      if(shape.indexOf(data.shape) !== -1 && color.indexOf(data.color) !== -1){
          filteredData.push(data)
      }
  return 0
  })
  if(shape.length === 5 && color.length === 6){
      filteredTitle = "All Items"
  }else if((color.length === 6 && shape.length > 1)||(color.length > 1 && shape.length === 5)){
    filteredTitle = "Multiple Items"
  }else if( shape.length ===5 && color.length === 1){
    filteredTitle = "All red Items"
  }else if( shape.length ===1 && color.length === 6){
    filteredTitle = `All ${shape[0]} item`
  } else if( shape.length > 1 && color.length === 1){
    filteredTitle = `Multiple ${color[0]} items`
  }else if( shape.length === 1 && color.length > 1){
    filteredTitle = `Multiple ${shape[0]} items`
  }else if( shape.length === 1 && color.length === 1){
    filteredTitle = `${color[0]}  ${shape[0]} item`
  }
    return {
        item:filteredData,
        filterName:filteredTitle
    }
}



const itemReducer = (state: itemState=initialState, action: Filter):any => {
    switch (action.type) {
        case "fetch-items": return state
        case "filter-items" : return filter(action.payload)
        default: return state
    }
}

export default itemReducer