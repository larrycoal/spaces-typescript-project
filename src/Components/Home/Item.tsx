import React from 'react';
import { useSelector } from 'react-redux';
import { State } from '../../Store/Reducer';



const Item: React.FC = () => {


    const { item, filterName } = useSelector((state: State) => state.items)

    const displayItems = (): any => {
        if(item.length === 0){
            return(
                <div className="not-found">
                    Nothing Found for this Filter
                </div>
            )
        }
        return item.map((item: any,i:number) => (
            <div key={i} className="item">
                <div
                    style={{ "background": item.color }}
                    className={item.shape}>

                </div>
            </div>
        ))
    }
    return (
        <div className="item-wrapper">
            <p>{filterName}</p>
            <div className="item-container">
                {displayItems()}
            </div>
        </div>
    );
};

export default Item;