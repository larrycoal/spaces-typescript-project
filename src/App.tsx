import React from 'react';
import { Route, Switch } from 'react-router';
import Header from './Components/Header';
import Home from './Components/Home/index';
import LoginPage from './Components/LoginPage';
import './Styles/styles.scss'

const App = () => {
  return (
    <>
      <Header />
      <Switch>
        <Route exact path="/home" component={Home} />
        <Route exact path="/" component={LoginPage} />
      </Switch>
    </>
  );
};

export default App;
