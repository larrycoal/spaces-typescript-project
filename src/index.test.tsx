import ReactDom from 'react-dom'
import React from "react"
import { Provider } from "react-redux"
import store from "./Store"
import { BrowserRouter } from "react-router-dom"
import App from "./App"
import userReducer from './Store/Reducer/userReducer'
import itemReducer from './Store/Reducer/itemReducer'
import data from './data.json'
import renderer from 'react-test-renderer'
import Item from './Components/Home/Item'
describe("App Component", () => {
    let container: HTMLDivElement

    beforeEach(() => {
        container = document.createElement('div')
        document.body.appendChild(container)
        const AppTest = (
            <Provider store={store}>
                <BrowserRouter>
                    <App />
                </BrowserRouter>
            </Provider>
        )
        ReactDom.render(AppTest, container)
    })
    afterEach(() => {
        document.body.removeChild(container);
        container.remove()
    })

    it("Login renders correctly", () => {
        const input = document.querySelectorAll('input')
        const button = document.querySelectorAll('button')
        expect(input).toHaveLength(2)
        expect(input[0].type).toBe("text")
        expect(input[1].type).toBe("password")
        expect(button).toHaveLength(2)
       
    })
    it("renders item correctly",()=>{
        const item = renderer.create(<Provider store={store}><Item/></Provider>).toJSON()
        expect(item).toMatchSnapshot()
    })
   
   
})
describe("reducer",()=>{
    const initialState = {
        item:data,
        filterName:"All items"
    }
    it("return user reducer",()=>{
    let newUser = userReducer({isLoggedIn:false},{type:"login",payload:true})
    expect(newUser).toEqual({isLoggedIn:true})
    })
    it("should return items",()=>{
        let item = itemReducer(initialState,{type:"fetch-items",payload:{shape:[],color:[]}})
        expect(item).toEqual(initialState)
    })
})