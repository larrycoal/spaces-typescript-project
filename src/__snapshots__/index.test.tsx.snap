// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders item correctly 1`] = `
<div
  className="item-wrapper"
>
  <p>
    All Items
  </p>
  <div
    className="item-container"
  >
    <div
      className="item"
    >
      <div
        className="oval"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="rectangle"
        style={
          Object {
            "background": "green",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="round"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="triangle"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="triangle"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "yellow",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="round"
        style={
          Object {
            "background": "blue",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="triangle"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "grey",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="rectangle"
        style={
          Object {
            "background": "blue",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="oval"
        style={
          Object {
            "background": "yellow",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "blue",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="oval"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="rectangle"
        style={
          Object {
            "background": "green",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="round"
        style={
          Object {
            "background": "blue",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="triangle"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="triangle"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "yellow",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="round"
        style={
          Object {
            "background": "black",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="triangle"
        style={
          Object {
            "background": "red",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "green",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="rectangle"
        style={
          Object {
            "background": "black",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="oval"
        style={
          Object {
            "background": "blue",
          }
        }
      />
    </div>
    <div
      className="item"
    >
      <div
        className="square"
        style={
          Object {
            "background": "yellow",
          }
        }
      />
    </div>
  </div>
</div>
`;
